device/iot-device-samples/send-receive-x509-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/SSLContextBuilder.java

```diff
@@ -0,0 +1,276 @@
+/*
+ *  Copyright (c) Microsoft. All rights reserved.
+ *  Licensed under the MIT license. See LICENSE file in the project root for full license information.
+ */
+
+package samples.com.microsoft.azure.sdk.iot;
+
+import org.apache.commons.codec.binary.Base64;
+
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.TrustManagerFactory;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.security.GeneralSecurityException;
+import java.security.Key;
+import java.security.KeyFactory;
+import java.security.KeyManagementException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.security.interfaces.RSAPrivateKey;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.UUID;
+
+/**
+ * Helper class that demonstrates how to build an SSLContext for x509 authentication from your public and private certificates,
+ * or how to build an SSLContext for SAS authentication from the default IoT Hub public certificates
+ */
+public class SSLContextBuilder
+{
+    private static final String SSL_CONTEXT_INSTANCE = "TLSv1.2";
+
+    private static final String CERTIFICATE_ALIAS = "cert-alias";
+    private static final String PRIVATE_KEY_ALIAS = "key-alias";
+
+    /**
+     * Create an SSLContext instance with the provided public and private keys that also trusts the base iot hub certificates
+     * @param publicKeyCertificateString the public key to use for x509 authentication. Does not need to include the default
+     *                                   Iot Hub trusted certificate as it will be added automatically
+     * @param privateKeyString The private key to use for x509 authentication
+     * @return The created SSLContext that uses the provided public key and private key
+     * @throws GeneralSecurityException If the certificate creation fails, or if the SSLContext creation using those certificates fails
+     * @throws IOException If the certificates cannot be read
+     */
+    public static SSLContext buildSSLContext(String publicKeyCertificateString, String privateKeyString) throws GeneralSecurityException, IOException
```

Comment:
With the 2.0.0 packages, all certificates will be read from the device's certificate store, even the certs for an x509 authenticated device/module identity. So I am assuming this sample will be modified once we have the code in place for reading certificates from the trusted certificates store, right?

Reply:
Kind of. We still need to demonstrate how to build an SSLContext for users who want to trust certificates other than the default ones installed on their machine in case they don't want to install the certs on the machine directly for any reason

Reply:
Notably, users who are doing x509 auth will still need to construct an SSLContext and pass it in for us to use. There is no way our SDK will be able to read private keys from the device directly without a user providing us the private key store and its password. I'll get more into that in the upcoming PR about reading certs from the device though

Reply:
Hmm, so this actually touches the point that first sparked this question in my mind. In neither of the sas-based or X509 certificates based flows does iot hub require the private key information, but the private key info seems to be necessary in building the ssl context. Is there no way to implement this in java which would get away from the requirement of the user having to supply the certificate private key to the SDK?

Just for comparison, the .NET SDK does not require the private key info to be sent with the certificate info that is sent to service. Instead, it is read directly from the certificate store by the TCP stack during TLS negotiation.

Reply:
I've remember this coming up as an issue mostly on linux systems where customers installed their certificates via openSsl and then restricted the access to only the tcp stack.

Reply:
So they don't even need to pass in a password for the key store that holds the private key? Interesting... I may need to do some more digging to see if we can mimic that in Java. For now though, I'll leave these samples as is

Reply:
> In neither of the sas-based or X509 certificates based flows does iot hub require the private key information

The client is expected to use the private key locally to encrypt the data that is sent over the wire. While the client doesn't send the private key to IoT Hub, it is still required

Reply:
> Instead, it (edit: is can be) read directly from the certificate store by the TCP stack during TLS negotiation.

A device can have multiple private keys installed though, right? How does the .NET SDK's TLS stack figure out which private key to use?

Reply:
Yes, the private key is required, for sure, what I mean is that it is required on a layer lower than the SDK deals with. I am not very confident on how the TLS stack figures out which private key to use, possibly through the common-name and thumbprint combination?